[READY] Implement new strategy to find the Python interpreter path

micbou opened the pull request:
See discussion in issue #1891 for details.

Implement a new strategy to find the Python interpreter path:
  - if specified, use `g:ycm_path_to_python_interpreter` option.
  - on UNIX platforms, use `sys.executable` as the path to Python interpreter;
  - on Windows, deduce it from `os.__file__` path (it should always be in the parent folder).

In all cases, check the version (2.6 or 2.7) of the Python interpreter path by running it.

This PR may break things. It needs thorough testing.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/1896)
<!-- Reviewable:end -->

micbou commented:
We should probably use [`sys.exec_prefix` instead of `os.__file__` on Windows](https://docs.python.org/2/library/multiprocessing.html#multiprocessing.set_executable). I'll update the PR.

Valloric commented:
This looks excellent. I'm far less concerned about this change now that I've seen it, but it's still a bit scary. Let's try it out on Windows, Mac and Linux.

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[python/ycm/paths.py, line 69 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1896#-K7nb_f64JUe4Tnz139-:-K7nb_f64JUe4Tnz1390:74405019) ([raw file](https://github.com/valloric/youcompleteme/blob/c7e32b3600020c65b39a0aba517a229609bcf23e/python/ycm/paths.py#L69)):*
A better name might be `IsPythonVersionCorrect`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1896#-:-K7nbmIXVgN9Kip1pQsR:-1079623987)*
<!-- Sent from Reviewable.io -->

vheon commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[python/ycm/paths.py, line 69 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1896#-K7nb_f64JUe4Tnz139-:-K7rjEHddTaAXu54HX47:-1756757540) ([raw file](https://github.com/valloric/youcompleteme/blob/c7e32b3600020c65b39a0aba517a229609bcf23e/python/ycm/paths.py#L69)):*
Was this comment on another line? because the `IsPythonVersionCorrect` is a couple of line down. :confused: 

---

*[python/ycm/tests/paths_test.py, line 32 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1896#-K7rioRC8DCR-umCSR3M:-K7rioRDFOCRwsJ9PMkx:1634980678) ([raw file](https://github.com/valloric/youcompleteme/blob/7378242e7da1c56fe6065685ede067f0efb59276/python/ycm/tests/paths_test.py#L32)):*
Can you add some negative tests? I mean if we test only for cases when the function return `True` we could have write:

```python
def EndsWithPython( path ):
  return True
```

and the tests will still be pass.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1896)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1896)*
<!-- Sent from Reviewable.io -->

micbou commented:
I was testing my code and I asked myself: since `sys.executable` can be a Vim path on some platforms, what happens if I replace directly the variable `python_interpreter` with gvim path? Let's find out.
Big mistake. `IsPythonVersionCorrect` started a new gvim instance which started in turn a new gvim instance which started... My screen was quickly covered with numerous gvim windows. Fortunately, I remembered one of my mentor's precept: *"when in disarray, `CTRL-C` you may"*. I frantically pressed the shortcut until the chain reaction was stopped. A sigh of relief when all this mess was cleaned up.

Moral of the story? We **really** don't want that to happen so we now check if the path ends with a Python name using a regex. I added tests for the regex.

I was also thinking maybe we should fallback to search Python in the `PATH` if this fails.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[python/ycm/paths.py, line 69 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1896#-K7nb_f64JUe4Tnz139-:-K7redKVHtVF2mEPdZdJ:972745840) ([raw file](https://github.com/valloric/youcompleteme/blob/c7e32b3600020c65b39a0aba517a229609bcf23e/python/ycm/paths.py#L69)):*
I just pushed a commit. It was `CheckPythonVersion` before.

---

*[python/ycm/tests/paths_test.py, line 32 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1896#-K7rioRC8DCR-umCSR3M:-K7rko6nepRkTXhfHcrA:-665593013) ([raw file](https://github.com/valloric/youcompleteme/blob/667b55af2477bc8e4a2584563a70ded4ee3c0e6e/python/ycm/tests/paths_test.py#L32)):*
Yes, I just added them.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1896#-:-K7rXx-ysL0-kFTZ158O:996663858)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[python/ycm/paths.py, line 69 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1896#-K7nb_f64JUe4Tnz139-:-K7rlt0Jqnsi4QNeynIX:-489270402) ([raw file](https://github.com/valloric/youcompleteme/blob/c7e32b3600020c65b39a0aba517a229609bcf23e/python/ycm/paths.py#L69)):*
Oh ok

---

*[python/ycm/tests/paths_test.py, line 39 \[r3\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1896#-K7rmceUo1H0ryQMD4OX:-K7rmceUo1H0ryQMD4OY:-1077890138) ([raw file](https://github.com/valloric/youcompleteme/blob/a80846e35dd733f094f8ced9daeacba58eeb4316/python/ycm/tests/paths_test.py#L39)):*
Can you add a `python3` failing test?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1896)*
<!-- Sent from Reviewable.io -->

micbou commented:
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[python/ycm/tests/paths_test.py, line 39 \[r3\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1896#-K7rmceUo1H0ryQMD4OX:-K7roW31CSURYPRvWPY7:-499146009) ([raw file](https://github.com/valloric/youcompleteme/blob/a80846e35dd733f094f8ced9daeacba58eeb4316/python/ycm/tests/paths_test.py#L39)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1896)*
<!-- Sent from Reviewable.io -->

vheon commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1896#-:-K7rqDYcZLJI1g-tBs0T:980077423)*
<!-- Sent from Reviewable.io -->

Valloric commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> from me as well.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[python/ycm/tests/paths_test.py, line 26 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1896#-K7sjqt7xuGI_wmsj5XL:-K7sjqt7xuGI_wmsj5XM:711543017) ([raw file](https://github.com/valloric/youcompleteme/blob/9a27c990db65650cc8505b40e485078e64747ec5/python/ycm/tests/paths_test.py#L26)):*
I was just about to write "you need to add tests for that regex" and you already did! :D

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1896#-:-K7sk3E8LcHO3YRoAIAt:-703173837)*
<!-- Sent from Reviewable.io -->

micbou commented:
Added Python search in PATH fallback. It should be even more robust this way.

---

Reviewed 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1896#-:-K8jDMtcF8iEOa3ri3mV:-424120785)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 1 of 1 files at r5.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1896)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Thanks for doing this! :)

@homu r+

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1896#-:-K8pwrAqlMBRfFpa7Ysl:2129009000)*
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit a0943d5 has been approved by `Valloric`

<!-- @homu r=Valloric a0943d5d31a7b80b22588bd0d30bbca6cd0798fc -->

homu commented:
:zap: Test exempted - [status](https://travis-ci.org/Valloric/YouCompleteMe/builds/104301179)
